br/pkg/lightning/backend/external/writer.go

```diff
@@ -173,7 +176,47 @@ func (b *WriterBuilder) Build(
 		kvStore:        nil,
 		onClose:        b.onClose,
 		closed:         false,
+		multiFileStats: make([]MultipleFilesStat, 1),
+		fileMinKeys:    make([]tidbkv.Key, 0, multiFileStatNum),
+		fileMaxKeys:    make([]tidbkv.Key, 0, multiFileStatNum),
 	}
+	ret.multiFileStats[0].Filenames = make([][2]string, 0, multiFileStatNum)
+	return ret
+}
+
+// MultipleFilesStat is the statistic information of multiple files (currently
+// every 500 files). It is used to estimate the data overlapping, and per-file
+// statistic information maybe too big to loaded into memory.
+type MultipleFilesStat struct {
+	MinKey            tidbkv.Key
+	MaxKey            tidbkv.Key
+	Filenames         [][2]string // [dataFile, statFile]
```

Comment:
Is one name enough?

Reply:
I have no strong opinion. If in future we won't have multiple data files share the same state file, I think we can remove it and provide a helper function to convert filenames.
And if you think the serialized metadata is too large, then it's a must to only keep filenames of data file. I'm not sure about it.